added smart contracts (#6762)
w3f/polkadot-wiki: docs/learn/learn-smart-contracts.md

---
title: Smart Contracts on Polkadot
description: Learn about Ethereum-compatible smart contracts n Polkadot
---

> Smart contracts are self-executing programs that run deterministically on a blockchain, enabling decentralized applications (dApps) without intermediaries. Polkadot natively supports smart contracts written in Ethereum-compatible languages like Solidity or Vyper, as well as Rust. Additionally, the underlying Polkadot Virtual Machine allows any smart contract that compiles to RISC-V to run on its network.

Reviewer: Not a big caveat, but iirc pvm is a slightly altered version of risc-v - but I think this statement should hold true regardless.

> Smart contracts are self-executing programs that run deterministically on a blockchain, enabling decentralized applications (dApps) without intermediaries. Polkadot natively supports smart contracts written in Ethereum-compatible languages like Solidity or Vyper, as well as Rust. Additionally, the underlying Polkadot Virtual Machine allows any smart contract that compiles to RISC-V to run on its network.

Reviewer: ```suggestion
Smart contracts are self-executing, sandboxed programs that run deterministically on a blockchain, enabling decentralized applications (dApps) without intermediaries. Polkadot natively supports smart contracts written in Ethereum-compatible languages like Solidity or Vyper, as well as Rust. Additionally, the underlying Polkadot Virtual Machine allows any smart contract that compiles to RISC-V to run on its network.
```

!!!info "Smart Contracts Documentation"
    For detailed guides on writing and deploying smart contracts on Polkadot, refer to the official [Polkadot Developer Docs](https://docs.polkadot.com/develop/). Note that PolkaVM is under active development.

## Smart Contracts vs. Rollups

Polkadot offers two primary ways to build decentralized applications:

- **Smart Contracts** - Deployed on existing blockchains, they are easier to develop but have limited flexibility in terms of governance and consensus.

> - **Parachains** - Independent blockchains connected to Polkadot, allowing full customization of business logic, consensus, and governance.

Reviewer: ```suggestion
- **Rollups** - Independent blockchains connected to and secured by Polkadot, allowing full customization of business logic, consensus, and governance. They are also known as _parachains_.
```

For a deeper comparison, read:  

> [Smart Contracts vs. Parachain Development](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/reference_docs/runtime_vs_smart_contract/index.html)

Reviewer: ```suggestion
[Smart Contracts vs. Rollup Development](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/reference_docs/runtime_vs_smart_contract/index.html)
```

## Smart Contracts on Polkadot

> Polkadot smart contracts will reside on the **Asset Hub**. When we talk about smart contracts on Polkadot, we're specifically referring to those deployed on the Asset Hub parachain, not the relay chain itself, whose main function stays securing parachain blocks and keeping the shared state. Smart contract functionality on the Hub is made possible thanks to the **pallet-revive**, an upgraded version of the original `pallet-contracts`. Unlike its predecessor (which used Wasm), `pallet-revive` executes contracts compiled to **Polkadot Virtual Machine (PVM)** bytecode.

Reviewer: IMO it would be nice to link to something, i.e., polkadot developer docs on pallet-revive and pallet-contracts (or something similar)

Reviewer 2: can do it on another PR

> ### Key Differences:

Reviewer: Perhaps this would be better put as a table?

Reviewer 2: Good suggestion! We can propose this later on.

- **pallet-contracts**: Uses **wasmi** (a WebAssembly interpreter).
- **pallet-revive**: Executes **PVM bytecode**, optimized for RISC-V architecture.

## Wasm vs. PVM

While WebAssembly (Wasm) is widely used in blockchain, it has limitations:

> 1. **Non-deterministic execution** - Compiled Wasm binaries are not always exactly the same, which is a problem for blockchain nodes that have to come to the exactly the same state. Although there are tools in the Polkadot ecosystem that solve this problem, this is still a fundamental limitation. 

Reviewer: ```suggestion
1. **Non-deterministic execution** - The slight variations between Wasm binaries present a hurdle for blockchain protocols, which must achieve an identical state for consensus. While tools within the Polkadot ecosystem exist to address this, it remains an underlying limitation.
```

2. **High computational cost** - Translating Wasm bytecode to machine code is computationally expensive. Wasm instructions are stack-based, while most hardware architecture is register-based with finite number of registers. Register allocation is a known [NP problem](https://en.wikipedia.org/wiki/Register_allocation#Common_problems_raised_in_register_allocation). 

> PVM bytecode solves both problems. 

Reviewer: ```suggestion
The Polkadot Virtual Machine (PVM) solves both of these problems by using the RISC-V architecture, which is a very common and well-supported standard.
```

###  RISC-V & PVM
 **RISC-V** is an **Instruction Set Architecture (ISA)**, not a bytecode, making it closer to native CPU execution.

PolkaVM (Parity's implementation of PVM) leverages RISC-V's register-based architecture, where operations use a finite set of CPU registers to pass arguments. This design mirrors modern hardware, enabling near 1:1 translation to native machine code. Unlike x86-64 (which suffers from register scarcity), PolkaVM's optimized register count simplifies compilation—reducing the complex register allocation problem to straightforward mapping. This is key to PolkaVM's rapid compilation speeds.

> Initially, smart contracts will be executed by the PolkaVM interpreter within the runtime itself. A later update will deliver a full PolkaVM JIT (Just in Time) compiler running inside the client with future optimizations planned.

Reviewer: Tiny nit: IMO this is okay to leave here, but maybe we should have some sort of comment / marker that reminds us to update things which refer to a future update, wdyt @filippoweb3?

Reviewer 2: yep this is important

## EVM vs. PVM

> It is important to note that while Polkadot will support Ethereum-compatible smart contracts (written in Solidity), it is incorrect to say that Polkadot is EVM-compatible, as no Ethereum Virtual machine implementayion is present in Polkadot's architecture including __pallet-revive__. There are miltiple reasons why Polkadot chose to deviate from traditional EVM support and decided to run smart contracts on PolkaVM. 

Reviewer: ```suggestion
It is important to note that while Polkadot will support Ethereum-compatible smart contracts (written in Solidity), it is incorrect to say that Polkadot is EVM-compatible, as no Ethereum Virtual machine implementations present in Polkadot's architecture including __pallet-revive__. There are multiple reasons why Polkadot chose to deviate from traditional EVM support and decided to run smart contracts on PolkaVM. 
```

> Just like Wasm EVM is stack-based, which results in slower execution time. EVM also uses a 256-bit word size, requiring all arithmetic operations to handle these large numbers. This design choice impacts performance, as each operation must be broken down into multiple native machine instructions. PolkaVM improves efficiency by using a 64-bit word size that matches modern hardware capabilities. However, when compiling Solidity contracts through YUL (via Revive), operations remain at 256-bit precision since YUL operates at too low a level to optimize integer types automatically. EVM code can only be interpreted, while PVM code can be interpreeted and compiled, which is optimal for larger multi-contract projects. EVm is als limited to Solidty, Vyper and other languages developed specifically for EVM, while PolkaVM supports any language that can be compiled to RISC-V, including Rust and C++. 

Reviewer 2: ```suggestion
Like Wasm, EVM is stack-based, resulting in slower execution time. EVM also uses a 256-bit word size, requiring all arithmetic operations to handle these large numbers. This design choice impacts performance, as each operation must be broken down into multiple native machine instructions. PolkaVM improves efficiency using a 64-bit word size that matches modern hardware capabilities. However, when compiling Solidity contracts through YUL (via Revive), operations remain at 256-bit precision since YUL operates at too low a level to optimize integer types automatically. EVM code can only be interpreted, while PVM code can be interpreted and compiled, which is optimal for larger multi-contract projects. EVM is also limited to Solidity, Vyper, and other languages explicitly developed for EVM, while PolkaVM supports any language that can be compiled to RISC-V, including Rust and C++. 
```

Some parachains support EVM smart contracts. 

## Revive: Solidity to PolkaVM Recompiler

**Revive** is the overarching project for compiling Solidity contracts to PVM bytecode. It includes:
- **resolc**: The Solidity-to-PVM compiler.
- **solc**: The Ethereum Solidity compiler (used internally).

> ### Compilation Flow:

Reviewer: Maybe we can have a mermaid.js diagram (not necessary!)
